packages/w3p/src/provider-detector.ts

```diff
@@ -0,0 +1,126 @@
+import { PROVIDER_CHECKS, PROVIDERS } from '@/enums'
+import { sleep } from '@/helpers'
+
+import { NearRawProvider } from './providers'
+import type {
+  EthereumProvider,
+  ProviderInstance,
+  RawProvider,
+  SolanaProvider,
+} from './types'
+
+declare global {
+  interface Window {
+    ethereum?: EthereumProvider
+    solana?: SolanaProvider
+    solflare?: SolanaProvider & {
+      isSolflare: boolean
+    }
+  }
+}
+
+export class ProviderDetector {
+  #providers: ProviderInstance[]
+  #rawProviders: RawProvider[]
+  #initiated = false
+
+  constructor() {
+    this.#providers = []
+    this.#rawProviders = []
+  }
+
+  public async init(): Promise<ProviderDetector> {
+    await sleep(500)
```

Comment:
mb there is a way to get rid of those sleeps? like subscribing to `ethereum#initialized` event in MM, maybe there are such events in other providers